lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+* Jar files installed in one of the
+  ${OPENSEARCH_HOME}/plugins
+  ${OPENSEARCH_HOME}/modules
+* Configuration passed to the main opensearch.yml, must be removed at plugin removal
+* Secrets stored in the keystore
+* The file: plugin.properties
+* And the security policy in: security.policy
+* Other plugins which it depends upon
+
+One last piece of the configuration is any index data uploaded to the cluster using
+OpenSearch APIs. That last bit of data must be done by inherinting the OpenSearchPlugin
+class and implementing the necessary extra logic.
+
+
+This file implements abstract methods and data that are common to every plugin during
+its lifecycle, including methods to manage configuration files, main processes (install,
+upgrade, uninstall, etc).
+
+The plugin lifecycle runs through the following steps:
+
+INSTALL > is_installed > is_enabled > needs_upgrade > upgrade > UNINSTALL
+
+The meaning of each step is, as follows:
+* is_installed: the installation happened correctly and the JAR files are set
+* is_enabled: all the configurations have been applied and restart is done, if needed
+* needs_upgrade: once the main OpenSearch is upgraded, the plugin needs to check if an
+                 upgrade is also needed or not.
+* upgrade: run the necessary actions to upgrade the plugin
+
+========================================================================================
+
+                             STEPS TO ADD A NEW PLUGIN
+
+========================================================================================
+
+
+For a new plugin, add the plugin to the list of "OpenSearchPluginsAvailable" available
+in opensearch_plugin_manager.py and override the abstract OpenSearchPlugin.
+
+Add a new configuration in the config.yaml with "plugin_" as prefix to its name.
+Add the corresponding config-name to the OpenSearchPluginsAvailable.
+
+If a given plugin depends on a relation, e.g. repository-s3, then add relation name
+as well. For example:
+    OpenSearchPluginsAvailable = {
+        ...
+        "opensearch-knn": {
+            "class": OpenSearchPlugin,
+            "config-name": "plugin_opensearch_knn",
+            "relation-name": ""
+        },
+    }
+"""
+
+import logging
+import os
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Union
+
+from charms.opensearch.v0.opensearch_exceptions import (
+    OpenSearchKeystoreError,
+    OpenSearchPluginError,
+)
+from jproperties import Properties
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, StatusBase
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchPlugin(ABC):
```

Comment:
Nice concept! 
We should make this class a declarative one, and offload all the operational work to the PluginManager.
We should then probably remove some methods and put them in the PluginManager and add a few methods such as: 
- `config()` => returns a dictionary of the config that must be added for the plugin to work. This will then be the job of the plugin manager to apply it on the `self.charm.opensearch_config`
- `install_source` => where is the plugin going to be picked up from for the plugin manager to locate it for the installation